spec.html

```diff
@@ -11059,6 +11060,26 @@ <h1>Examples of Automatic Semicolon Insertion</h1>
       <p>In the circumstance that an assignment statement must begin with a left parenthesis, it is a good idea for the programmer to provide an explicit semicolon at the end of the preceding statement rather than to rely on automatic semicolon insertion.</p>
     </emu-clause>
   </emu-clause>
+  <emu-clause id="sec-hazards-of-automatic-semicolon-insertion">
+    <h1>Hazards of Automatic Semicolon Insertion</h1>
+    <em>This section is non-normative.</em>
+    <p>ECMAScript programs can be written in a style with very few semicolons by relying on automatic semicolon insertion. However, as described above, semicolons are not inserted at every newline, and some of these cases which do not have an automatically inserted semicolon can be counter-intuitive or confusing.</p>
+
+    <p>As new syntactic features are added to ECMAScript, additional cases requiring explicit semicolons emerge over time. As such, consistently explicit semicolon use is recommended.</p>
```

Comment:
> As such, consistently explicit semicolon use is recommended.

According to [the hazards I described on the previous comment](https://github.com/tc39/ecma262/pull/1062/files#r161150751), consistent use of explicit semicolons will not solve those cases.

Since these hazards apply to code of all styles. I’d like to suggest recommending users of the language to “understand these ASI hazards” (instead of just using explicit semicolons, but unaware of the aforementioned hazards), or, of course, use a tool that warns them about these hazards.

However, if the sentence is unrelated to the ASI hazards, but related to the fact that “additional cases requiring explicit semicolons emerge over time,” then it should be made clear that the recommendation of using explicit semicolons are not about “safety” or “reducing confusion” (as many articles have proved otherwise), but rather “consistency.” Maybe it can be reworded like this to make the intent more clear:

> As future syntax constructs may require explicit semicolons, to maintain a consistent code style, explicit semicolon use is recommended.

Reply:
> As future syntax constructs may require explicit semicolons, to maintain a consistent code style as the language evolves, explicit semicolon use is recommended

These words are still implying semicolon-less style brings inconsistent. In my opinion, semicolon-less style only insert semicolon when really needed, these necessary semicolons are not inconsistent, just exposes necessary evil which derived from the first day of  language. So just replace the last sentence "As such, consistently explicit semicolon use is recommended." to 

"As such, using up-to-date linter is recommended."

Reply:
I guess, as a part of language it is essential to be explicit in certain areas. Being implicit adds complexity and learning curve of the language. Things like this and `;` and others should be handled and made more explicit. Such that it is much easier for the programmers to spend time on something amazing rather than worrying about the language implicitness. 👍  🎉

Reply:
rephrased a bit